moped-editor/src/views/projects/projectView/ProjectTeam/ProjectTeamTable.js

```diff
@@ -0,0 +1,564 @@
+      notes,
+      moped_proj_personnel_roles,
+    },
+    projectId: project_id,
+  }) => {
+    const payload = { notes, project_id, user_id };
+    const personnelRoles = moped_proj_personnel_roles.map((roles) => ({
+      project_role_id: roles.project_role_id,
+    }));
+
+    payload.moped_proj_personnel_roles = { data: personnelRoles };
+    return payload;
+  };
+
+  /**
+   * Construct a moped_project_personnel object that can be passed to an update mutation
+   * @param {Object} newData - a table row object with { moped_user, notes, roleIds }
+   * @param {integer} projectId - the project ID
+   * @return {Object} a moped_project_personnel object: { user_id, notes } <- observe that `moped_proj_personnel_roles`
+   *  is handled separately
+   */
+  const getEditPersonnelPayload = (newData) => {
+    const {
+      moped_user: { user_id },
+      notes,
+    } = newData;
+    return { user_id, notes };
+  };
+
+  /**
+   * Constructs payload objects for adding and removing moped_proj_personnel_roles
+   * @param {Object} newData - a table row object with the new values
+   * @param {Object} oldData - a table row object with the old values
+   * @return {[[Object], [Int]]} - an array of new personnel role objects, and an array of existing
+   *  personnel role objects to delete
+   */
+  const getEditRolesPayload = (newData, oldData) => {
+    const { project_personnel_id } = oldData;
+
+    // get an array of moped_proj_personnel_roles IDs to delete
+    const projRoleIdsToDelete = oldData.moped_proj_personnel_roles
+      .filter((projRole) => {
+        const roleId = projRole.moped_project_role.project_role_id;
+        return !newData.roleIds.includes(roleId);
+      })
+      .map((projRole) => projRole.id);
+
+    // construct an array of new moped_proj_personnel_roles objects
+    const existingRoleIds = oldData.moped_proj_personnel_roles.map(
+      ({ moped_project_role }) => moped_project_role.project_role_id
+    );
+    const roleIdsToAdd = newData.roleIds.filter(
+      (roleId) => !existingRoleIds.includes(roleId)
+    );
+    const rolesToAddPayload = roleIdsToAdd.map((newRoleId) => ({
+      project_personnel_id,
+      project_role_id: newRoleId,
+    }));
+    return [rolesToAddPayload, projRoleIdsToDelete];
+  };
+
+  const onClickAddTeamMember = () => {
+    const id = uuidv4();
+    setRows((oldRows) => [
+      {
+        id,
+        name: null,
+        moped_workgroup: null,
+        moped_proj_personnel_roles: [],
+        notes: null,
+        isNew: true,
+        roleIds: [],
+        project_personnel_id: id,
+      },
+      ...oldRows,
+    ]);
+    setRowModesModel((oldModel) => ({
+      ...oldModel,
+      [id]: { mode: GridRowModes.Edit, fieldToFocus: "project_personnel_id" },
+    }));
+  };
+
+  const handleEditClick = useCallback(
+    (id) => () => {
+      setRowModesModel({ ...rowModesModel, [id]: { mode: GridRowModes.Edit } });
+    },
+    [rowModesModel]
+  );
```

Comment:
Do you think we should blank out the value in the "Workgroup" column when you go into edit mode, or maybe when you change a person's name while in edit mode?

As it is, you have this awkward Intermezzo moment after you change a name and before you save your change when the name and the workgroup do not match. 

It makes me feel like I am making a mistake or somehow causing the "correctness" level of the situation to go down. I want to fix it, and I can't, and I don't want to hit save to lock in the degraded state.

Reply:
thanks for catching this Frank! this is very similar to the milestone and related phase form behavior that @chiaberry recently figured out. ✨

I created an issue so we can do this later. https://github.com/cityofaustin/atd-data-tech/issues/20328